numpy/core/records.py

```diff
@@ -423,6 +423,11 @@ def __new__(subtype, shape, dtype=None, buf=None, offset=0, strides=None,
                                       strides=strides, order=order)
         return self
 
+    def __array_finalize__(self, obj):
+        if type(obj) is not type(self):
+            # invoke __setattr__
+            self.dtype = self.dtype
```

Comment:
I expanded the comment, but I still have to use `self.dtype` since the condition is "self.dtype is **not** obj.dtype" ;)